### ISSUE TITLE
One equivalent query can help improve performance hundred of times

### ISSUE THREAD

**Issue by YuanchengJiang** [neo4j/neo4j#12973]
Neo4j version: 4.4.12
Operating system: Ubuntu 20.04
API/Driver: Cypher
Dataset: Pole(Crime analysis), https://github.com/neo4j-graph-examples/pole

Query 1:
```
 MATCH p=(n1)--()<-[r1]-()--(r2:PostCode)--(n2)<--()-->()-->()<-[r3:POSTCODE_IN_AREA]-(n1) RETURN n2 as X SKIP 6 LIMIT 2;
```
It is a common query and takes more than one hour to finish on my PC.

Query 2:
```
CALL { MATCH p=(n1)--()<-[r1]-()--(r2:PostCode)--(n2)<--()-->()-->()<-[r3:POSTCODE_IN_AREA]-(n                                                                                                   
1) RETURN n2 as X SKIP 6 LIMIT 2 } RETURN X;
```
This is an equivalent query to Query 1 which is wrapped by CALL. However, it is done within a few seconds.

Query 1 Profiles:
```
neo4j@neo4j> PROFILE MATCH p=(n1)--()<-[r1]-()--(r2:PostCode)--(n2)<--()-->()-->()<-[r3:POSTCODE_IN_AREA]-(n1) RETURN n2 as X SKIP 6 LIMIT 2;                                                                                                             
+---+                                                         
| X |                                                         
+---+                                                         
+---+                                                         

+-----------------------------------------------------------------------------------------------------------------+          
| Plan      | Statement   | Version      | Planner | Runtime       | Time    | DbHits     | Rows | Memory (Bytes) |          
+-----------------------------------------------------------------------------------------------------------------+          
| "PROFILE" | "READ_ONLY" | "CYPHER 4.4" | "COST"  | "INTERPRETED" | 4642011 | 6272095849 | 0    | 4463896        |          
+-----------------------------------------------------------------------------------------------------------------+          


+------------------------+------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| Operator               | Details                                                                                              | Estimated Rows | Rows       | DB Hits    | Memory (Bytes) | Page Cache Hits/Misses |                                    
+------------------------+------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +ProduceResults@neo4j  | X                                                                                                    |              2 |          0 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Projection@neo4j      | n2 AS X                                                                                              |              2 |          0 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Skip@neo4j            | $autoint_0                                                                                           |              2 |          0 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Limit@neo4j           | 2 + $autoint_0                                                                                       |              3 |          0 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Filter@neo4j          | not anon_0 = r1 AND not anon_0 = anon_7 AND not anon_0 = anon_3 AND not anon_0 = anon_4 AND not anon |              3 |          0 |          0 |                |                    0/0 |                                    
| |                      | _0 = anon_5 AND not anon_0 = anon_9 AND not anon_0 = r3                                              |                |            |            |                |                        |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(Into)@neo4j    | (n1)-[anon_0]-(anon_1)                                                                               |              3 |  662022320 |      66331 |        4463816 |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Filter@neo4j          | not anon_5 = r1 AND not r1 = r3 AND not anon_4 = r1 AND not anon_3 = r1 AND not anon_7 = r1 AND not  |          57550 |  911418314 |          0 |                |                    0/0 |                                    
| |                      | anon_9 = r1                                                                                          |                |            |            |                |                        |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(All)@neo4j     | (anon_2)-[r1]->(anon_1)                                                                              |          61127 | 1822836628 | 2734254942 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Filter@neo4j          | not anon_3 = anon_4 AND not anon_3 = anon_9 AND not anon_3 = anon_7 AND not anon_3 = anon_5 AND not  |          35531 |  911418314 |          0 |                |                    0/0 |                                    
| |                      | anon_3 = r3                                                                                          |                |            |            |                |                        |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(All)@neo4j     | (r2)-[anon_3]-(anon_2)                                                                               |          37362 | 1774158950 | 2636899586 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Filter@neo4j          | not anon_9 = r3 AND not anon_7 = r3 AND not anon_5 = r3 AND not anon_4 = r3                          |          18226 |  862740636 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(All)@neo4j     | (anon_10)<-[r3:POSTCODE_IN_AREA]-(n1)                                                                |          18974 |  868371898 |  871194640 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Filter@neo4j          | not anon_5 = anon_9 AND not anon_4 = anon_9 AND not anon_7 = anon_9                                  |          82228 |    2822742 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(All)@neo4j     | (anon_8)-[anon_9]->(anon_10)                                                                         |          84745 |    2835822 |    5695186 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Filter@neo4j          | not anon_5 = anon_7 AND not anon_4 = anon_7                                                          |          49259 |    2859364 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(All)@neo4j     | (anon_6)-[anon_7]->(anon_8)                                                                          |          50259 |    9867701 |   16876038 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Filter@neo4j          | not anon_4 = anon_5                                                                                  |          29214 |    7008337 |          0 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(All)@neo4j     | (n2)<-[anon_5]-(anon_6)                                                                              |          29509 |    7022533 |    7051633 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +Expand(All)@neo4j     | (r2)-[anon_4]-(n2)                                                                                   |          17152 |      29100 |      43296 |                |                    0/0 |                                    
| |                      +------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    
| +NodeByLabelScan@neo4j | r2:PostCode                                                                                          |           8368 |      14196 |      14197 |                |                    0/0 |                                    
+------------------------+------------------------------------------------------------------------------------------------------+----------------+------------+------------+----------------+------------------------+                                    

0 rows                                                        
ready to start consuming query after 197 ms, results consumed after another 4641814 ms              
```

Query 2 Profiles:
```
neo4j@neo4j> PROFILE CALL { MATCH p=(n1)--()<-[r1]-()--(r2:PostCode)--(n2)<--()-->()-->()<-[r3:POSTCODE_IN_AREA]-(n                                                                                                   
1) RETURN n2 as X SKIP 6 LIMIT 2 } RETURN X;                                                                                 
+---+                                                                                                                                                                                                                 
| X |                                                                                                                                                                                                                 
+---+                                                                                                                                                                                                                 
+---+                                                                                                                                                                                                                 
                                                                                                                                                                                                                      
+------------------------------------------------------------------------------------------------------------+                                                                                                        
| Plan      | Statement   | Version      | Planner | Runtime       | Time | DbHits   | Rows | Memory (Bytes) |                                                                                                        
+------------------------------------------------------------------------------------------------------------+                                                                                                        
| "PROFILE" | "READ_ONLY" | "CYPHER 4.4" | "COST"  | "INTERPRETED" | 5984 | 14829046 | 0    | 64             |                                                                                                        
+------------------------------------------------------------------------------------------------------------+                                                                                                        
                                                                                                                                                                                                                      
                                                                                                                                                                                                                      
+--------------------------+------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| Operator                 | Details                                                                                              | Estimated Rows | Rows    | DB Hits | Page Cache Hits/Misses |                     
+--------------------------+------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +ProduceResults@neo4j    | X                                                                                                    |              0 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Projection@neo4j        | n2 AS X                                                                                              |              0 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Skip@neo4j              | 6                                                                                                    |              0 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Limit@neo4j             | 2 + 6                                                                                                |              5 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Filter@neo4j            | not anon_4 = anon_9 AND not anon_3 = anon_4 AND not anon_5 = r1 AND not anon_7 = r3 AND not anon_0 = |              5 |       0 |       0 |                    0/0 |                     
| |                        |  anon_7 AND not anon_4 = r1 AND not anon_5 = r3 AND not anon_0 = anon_4 AND not anon_0 = anon_5 AND  |                |         |         |                        |                     
| |                        | not anon_7 = r1 AND not anon_3 = anon_7 AND not anon_5 = anon_9 AND not anon_7 = anon_9 AND not anon |                |         |         |                        |                     
| |                        | _4 = r3 AND not anon_3 = anon_5                                                                      |                |         |         |                        |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +NodeHashJoin@neo4j      | anon_8, r2                                                                                           |              6 |       0 |       0 |                    0/0 |                     
| |\                       +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| | +Filter@neo4j          | not anon_5 = anon_7 AND not anon_4 = anon_7                                                          |          85267 |       0 |       0 |                    0/0 |                     
| | |                      +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| | +Expand(All)@neo4j     | (anon_6)-[anon_7]->(anon_8)                                                                          |          85267 |       0 |       0 |                    0/0 |                     
| | |                      +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| | +Filter@neo4j          | not anon_4 = anon_5                                                                                  |          49563 |       0 |       0 |                    0/0 |                     
| | |                      +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| | +Expand(All)@neo4j     | (n2)<-[anon_5]-(anon_6)                                                                              |          50063 |       0 |       0 |                    0/0 |                     
| | |                      +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| | +Expand(All)@neo4j     | (r2)-[anon_4]-(n2)                                                                                   |          29100 |       0 |       0 |                    0/0 |                     
| | |                      +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| | +NodeByLabelScan@neo4j | r2:PostCode                                                                                          |          14196 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Filter@neo4j            | not anon_9 = r3 AND not anon_3 = anon_9 AND not anon_0 = anon_9 AND not anon_9 = r1                  |          61844 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Expand(All)@neo4j       | (anon_10)<-[anon_9]-(anon_8)                                                                         |          64381 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Filter@neo4j            | not anon_0 = r3 AND not anon_3 = r3 AND not r1 = r3                                                  |          37422 |       0 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Expand(All)@neo4j       | (n1)-[r3:POSTCODE_IN_AREA]->(anon_10)                                                                |          38568 | 2828711 | 8755774 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Filter@neo4j            | not anon_0 = anon_3 AND not anon_0 = r1                                                              |         167140 | 5927063 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Expand(All)@neo4j       | (anon_1)-[anon_0]-(n1)                                                                               |         170534 | 5941967 | 5956871 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Filter@neo4j            | not anon_3 = r1                                                                                      |          49563 |   14904 |       0 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +Expand(All)@neo4j       | (anon_2)-[r1]->(anon_1)                                                                              |          50063 |   29808 |   58908 |                    0/0 |                     
| |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                                                         
| +Expand(All)@neo4j       | (r2)-[anon_3]-(anon_2)                                                                               |          29100 |   29100 |   43296 |                    0/0 |                                                         | |                        +------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                     
| +NodeByLabelScan@neo4j   | r2:PostCode                                                                                          |          14196 |   14196 |   14197 |                    0/0 |                     
+--------------------------+------------------------------------------------------------------------------------------------------+----------------+---------+---------+------------------------+                                                         

0 rows                                                        
ready to start consuming query after 8 ms, results consumed after another 5976 ms  
```

**Comment by HannesSandberg** [comment]
Thanks! We will investigate this and report back to you.

**Comment by sherfert** [comment]
Hello @YuanchengJiang 

We found out why the subquery version with `CALL` performed so much better.
A fix will come with the next neo4j release: 5.4. We will not be rolling out this fix to the 4.4 series.

Please also note that the planner doesn't really have a lot to work with in this query, almost no labels nor types or anything it can use to get useful estimations. These kind of queries are very vulnerable to bad estimates leading to poor performance.
